# Eugene's Integral Computations - Galerkin Method of Moments

## Import Packages

In [1]:
using StaticArrays 
using LinearAlgebra # provides the function norm 
using HCubature # provides adaptive numerical integration 

using Test # provides testing functionality 
using Plots 

const Point3D = SVector{3,Float64};

## Section 1: Introduction

More later.

For information on decomposin elements into faces, edges and points, see [wiki polygon_mesh](https://en.wikipedia.org/wiki/Polygon_mesh) 
## Section 2: Mathematical Preliminaries 

Integration by parts for grad-div equations. Assume $\mathbf{V}(\mathbf{r})$ and $\mathbf{W}(\mathbf{r})$ to be two vector fields on $\Omega \subset \mathbb{R}^3$. Then 

$$
\text{div} \left[ \mathbf{V} \,  \text{div} \mathbf{W} \right] = 
\text{div} \mathbf{V} \, \text{div} \mathbf{W} + \mathbf{V} \cdot \text{grad} \left( \text{div} \mathbf{W} \right) \, . 
$$

## Section 3: Magnetic Method of Moments 
 
**Problem Formulation** Assume given a metalic object with magnetic susceptibility $\chi_{mag}$ (assumed constant here, value?) and with volume $\Omega \subset \mathbb{R}^3$ (typical dimensions?) placed in a constant external magnetic field in the $x$-direction $\mathbf{H}_0 = \left( H_0, 0,0\right)$ (value?, units?). Then the object is magnitized, meaning that the magnetization vector $\mathbf{M}(\mathbf{r}) = \left( M_x(\mathbf{r}), M_y(\mathbf{r}), M_z(\mathbf{r})\right)$ is non-zero inside $\Omega$. We wish to compute $\mathbf{M}(\mathbf{r})$ by solving the linear vector-valued grad-div volume integro-differential equation with singular kernel given by  

$$
\mathbf{A} \left[ \mathbf{M}(\mathbf{r}) \right] = \mathbf{H}_0
$$

where 

$$
\mathbf{A}\left[ \mathbf{M}(\mathbf{r}) \right] = \frac{1}{\chi_{mag}} \mathbf{M}(\mathbf{r}) - 
\text{grad}_{\mathbf{r}} \, \text{div}_{\mathbf{r}} 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' = 
\frac{1}{\chi_{mag}} \mathbf{M}(\mathbf{r}) - 
\nabla_{\mathbf{r}} \, \nabla_{\mathbf{r}} \cdot 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega'
$$

The magnetization $\mathbf{M}$ remains zero outside $\Omega$, i.e., on $\mathbb{R}^3 \setminus \Omega$ (the equivalent of homogeneous Dirichlet boundary conditions). The integro-differential equation above is a system of three coupled equations for the three components of the magnetization $\mathbf{M}(\mathbf{r})$. 

For future referencde, we write the equation in more explicit form. The first of the three scalae equations is 

$$
\frac{1}{\chi_{mag}} M_x(\mathbf{r}) - 
\frac{\partial}{\partial x} \, \nabla_{\mathbf{r}} \cdot 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' = H_{0,x} \, , 
$$

where 

$$
\nabla_{\mathbf{r}} \cdot 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' = 
\frac{\partial}{\partial x} \int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' + 
\frac{\partial}{\partial y} \int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' 
+
\frac{\partial}{\partial z} \int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \, . 
$$ 

(Information on the regularity of the computed magnetization would be valuable to have.)

**Two formulations** We distinguish a **special-case formulation** and a **general formulation**. In the special-case formulation, the object in considered to be thin. This implies that the $z$-dependence of the components of the magnetization vector $\mathbf{M}$ is neglected. The components of the magnetization vector can be expressed as $M_x = M_x(x,y)$, $M_y=M_y(x,y)$ and $M_z = M_z(x,y)$. The kernel remains dependent on $z$ and $z'$ resulting in 6D integrals. In the general formulation, components of the magnetization vector are . Both formulation are developed for **linear** basis or shape function.

## Section 4: Galerkin Method    

**Continuous Formulation** A continuous formulation of the Galerkin would be valuable to developed. Below, we develop a spatially discrete variant only.  

**Mesh generation** We assume a conformal mesh on $\Omega$ to be denoted by $\Omega^h$. In the special-case formulation, the mesh consists of prisms generated by extruding the triangular mesh in the $z$-direction. Each element has 2 triangular and 3 quadrilateral faces. In the general formulation, the mesh consists of tetrahedra. Each element consists of 4 triangular faces. We denote the number of nodes, edges, facets and elements of $\Omega^h$ by $N_n$, $N_{ed}$, $N_f$ and $N_e$, respectively. Assume the elements to be denoted by $\left\{ P_{\alpha} | 1 \leq \alpha \leq N_e \right\}$. Assume $g(\mathbf{r})$ and $h(\mathbf{r})$ to be functions on $\Omega$ such that 

$$
g(\mathbf{r}) = \int_{\Omega} \frac{h(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \, . 
$$

Then by decomposing $\Omega$ into a set of volumes $P_{\alpha}$, we find that 

$$
\begin{eqnarray}
\int_{\Omega} g(\mathbf{r}) \, d\Omega = 
\sum_{\alpha=1}^{N_e}  \int_{P_{\alpha}} g(\mathbf{r}) \, d\Omega 
& = & \sum_{\alpha,\beta=1}^{N_e}  \int_{P_{\alpha}} \int_{P_{\beta}} \frac{h(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \, d\Omega \\ 
& = & \sum_{\alpha,\beta=1}^{N_e}  \int_{\mathbb{R}^3} \int_{\mathbb{R}^3} \frac{h(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, \mathbb{1}_{P_{\beta}} \, d\Omega' \, \mathbb{1}_{P_{\alpha}} \, d\Omega
\end{eqnarray}
$$ 

We assume that the information to decompose an element $P_{\alpha}$ into a set of facets, a facet into a set of edges and an edge into a set of points to be available (representation of the mesh $\Omega^h$ as a directed a-cyclic graph (DAG) with corresponding operations to find parent nodes and child nodes. See e.g. [wikipedia entry on polygon mesh](https://en.wikipedia.org/wiki/Polygon_mesh)).  

**Set of Nodal Basis Functions** The mesh $\Omega^h$ allows to define a set of nodal P1 linear Lagrange basis functions $\{ \phi_i(\mathbf{r}) | 1 \leq i \leq N _n \}$ (see classical finite elements method). The functions $\phi_i(\mathbf{r})$ are scalar functions. To be able to cast a vector-valued partial differential equation in weak or variational form, we define $3 \, N_e$ vector-valued shape functions $\boldsymbol{\phi}_{3\,i-2} = \left( \phi_i, 0,0\right)$, 
$\boldsymbol{\phi}_{3i-1} = \left( 0, \phi_i, 0\right)$ and 
$\boldsymbol{\phi}_{3i} = \left( 0, 0, \phi_i \right)$ for $1 \leq i \leq N_n$.  

**Expansion of Magnetization Components** The numerical approximation of the magnetization components can be expanded in as 

\begin{eqnarray}
M_x(\mathbf{r}) = \sum_{i=1}^{N_n} c_i \, \phi_i(\mathbf{r}) \\ 
M_y(\mathbf{r}) = \sum_{i=1}^{N_n} d_i \, \phi_i(\mathbf{r}) \\
M_z(\mathbf{r}) = \sum_{i=1}^{N_n} e_i \, \phi_i(\mathbf{r}) 
\end{eqnarray}

Note that components $M_x(\mathbf{r})$, $M_y(\mathbf{r})$ and $M_z(\mathbf{r})$ are each expanded in the same basis with proper expansion coefficients. In case of linear Lagrange shape functions on tetrahedra, the element $P_{\alpha}$ has 4 local degrees of freediom per components, thus in total $3*4=12$ local degrees of freedom. Mapping from local to global degrees of freedom. The expansion coefficients $c_i$, $d_i$ and $e_i$ can stacked into a global vector $\mathbf{u}$ of dimension $3 \, N_n$ where $\mathbf{u}^T = \begin{pmatrix} \mathbf{c}^T \mathbf{d}^T \mathbf{e}^T \end{pmatrix}$. This global vector is the solution of the Galerkin-MoM linear system 

$$
\underline{\underline{A}} \, \mathbf{u} = \mathbf{b}
$$

where the matrix $A$ is a dense $3 \, N_n$-by-$3 \, N_n$ matrix consisting of a stiffness part $A^{(2)}$ and a mass matrix part $A^{(1)}$.  

**Discrete Weak Form**

$$
\int_{\Omega} \mathbf{A} \left[ \mathbf{M}(\mathbf{r}) \right] \cdot \boldsymbol{\phi}_k \, d\Omega = 
\int_{\Omega} \mathbf{H}_0 \cdot \boldsymbol{\phi}_k \, d\Omega 
\hspace{1cm} \forall 1 \leq k \leq N_e 
$$ 

or more explicitly 

$$
\frac{1}{\chi_{mag}} \int_{\Omega} \mathbf{M}(\mathbf{r}) \cdot \boldsymbol{\phi}_k \, d\Omega + 
\int_{\Omega} \left[ \nabla_{\mathbf{r}} \, \nabla_{\mathbf{r}} \cdot 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \right] \cdot \boldsymbol{\phi}_k \, d\Omega = \int_{\Omega} \mathbf{H}_0 \cdot \boldsymbol{\phi}_k \, d\Omega
$$ 

and thus after integration by parts 

$$
\frac{1}{\chi_{mag}} \int_{\Omega} \mathbf{M}(\mathbf{r}) \cdot \boldsymbol{\phi}_k \, d\Omega - 
\int_{\Omega} \left[ \nabla_{\mathbf{r}} \cdot \boldsymbol{\phi}_k \right]
\left[\nabla_{\mathbf{r}} \cdot 
\int_{\Omega} \frac{\mathbf{M}(\mathbf{r}')}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \right] d\Omega = \int_{\Omega} \mathbf{H}_0 \cdot \boldsymbol{\phi}_k \, d\Omega
$$

resulting in $3 \, N_e$ equations for the $3 \, N_e$ components of the vector of expansion coefficients $\mathbf{u}$. The mass matrix and load vector can be assembled by a loop over elements as in classical Galerlin FEM. The stiffness matrix requires dedicated attention due to the integral term.   

## Section 5: Element-by-element Assembly of Stiffness Matrix

**Local Representation**

Assume that $P_{\alpha}$ has $N_e^{\alpha}$ nodes. Assume that $\phi_k({\mathbf r}) = a_k x + b_k y + c_k y + d_k$ is a (scalar) basis function for $1 \leq k \leq N_e^{\alpha}$.  

**Bilinear Form**

$$
\underline{\underline{A}}^{(2)} = \sum_{\alpha} \underline{\underline{A}}_{\alpha}^{(2)} = \sum_{\alpha\beta} \underline{\underline{A}}_{\alpha\beta}^{(2)} \text{ of size } 3 \, N_n \text{ by } 3 \, N_n 
$$

where the sum over $\alpha$ (destination) and $\beta$ (source) discretize the integral over $\mathbf{r}$ and $\mathbf{r}'$, respectively. In case that $\alpha = \beta$, self-interaction terms are considered. 

$$
\underline{\underline{A}}_{loc,\alpha\beta}^{(2)} = 
\begin{pmatrix} 
A_{\alpha\beta}^{xx} & A_{\alpha\beta}^{xy} &  A_{\alpha\beta}^{xz} \\
A_{\alpha\beta}^{yx} & A_{\alpha\beta}^{yy} &  A_{\alpha\beta}^{yz} \\ 
A_{\alpha\beta}^{zx} & A_{\alpha\beta}^{zy} &  A_{\alpha\beta}^{zz}
\end{pmatrix}
\text{ of size } 12 \text{ by } 12
$$

where (row determined by the expansion of the magnetization, columns determined by the test function). 

$$
A_{\alpha\beta}^{xx} = 
\begin{pmatrix} 
A_{\alpha\beta}^{xx,11} & A_{\alpha\beta}^{xx,12} & A_{\alpha\beta}^{xx,13} & A_{\alpha\beta}^{xx,14} \\
A_{\alpha\beta}^{xx,21} & A_{\alpha\beta}^{xx,22} & A_{\alpha\beta}^{xx,23} & A_{\alpha\beta}^{xx,24} \\
A_{\alpha\beta}^{xx,31} & A_{\alpha\beta}^{xx,32} & A_{\alpha\beta}^{xx,33} & A_{\alpha\beta}^{xx,34} \\
A_{\alpha\beta}^{xx,41} & A_{\alpha\beta}^{xx,42} & A_{\alpha\beta}^{xx,43} & A_{\alpha\beta}^{xx,44} 
\end{pmatrix}
\text{ of size } 4\text{ by } 4 
$$

where 

$$
\begin{eqnarray}
A_{\alpha\beta}^{xx,11} & =  & - \int_{P_{\alpha}} \frac{\partial \phi_1({\mathbf r})}{\partial x} \, \frac{\partial}{\partial x} \int_{P_{\beta}} \frac{\phi_1({\mathbf r})}{\|\mathbf{r}' - \mathbf{r} \|} \, d\Omega' \, d\Omega \\
& =  & - \int_{\mathbb{R}^3} \frac{\partial  \phi_1({\mathbf r})}{\partial x} \, \frac{\partial}{\partial x} \int_{\mathbb{R}^3} \frac{\phi_1({\mathbf r})}{\|\mathbf{r}' - \mathbf{r} \|} \, \mathbb{1}_{P_{\beta}} \, d\Omega' \, \mathbb{1}_{P_{\alpha}} \,d\Omega
\end{eqnarray}
$$ 

and similar for other components. Introduce indicator functions on $P_{\alpha}$ and $P_{\beta}$ and integrate over entire $\mathbb{R}^3$. This will allow to use the divergence theorem and the fact that the basis functions centered on $P_{\alpha}$ and $P_{\beta}$ are zero sufficiently far away from $P_{\alpha}$ and $P_{\beta}$.     

**Reduce 6D integrals to 4D integrals**

1. reduction of inner integral from 3D to 2D: interchange order of differentiation and integration, obtain two terms, apply Euler on one term;
2. reduction of outer integral from 3D to 2D: derivate of linear shape function is a constant. Apply Green-Gauss divergence theorem; 

## Section 5: Other Information 

### How to perform quadrature over triangular facet 

In two steps: 
1. first coordinate transformation from $(x,y,z)$ to $(\xi,\eta)$ (details missing); 
2. second coordinate transformation from $0 \leq \eta \leq \xi$ to $0 \leq \xi \, u \leq 1$; 

### Bottom-up Approach 

1. implement the testing by comparing symbolic and numerical evaluation of the integrals; 

### Top-down Approach 

#### Two facets test cases 
1. compare analytical and numerical computation

#### Two tetrahedra test case 
1. assume two tetrahedra to be given;
1. two identical tetrahedra $P_{\alpha}$ and $P_{\alpha}$
2. two neighbouring tetrahedra $P_{\alpha}$ and $P_{\alpha'}$ sharing a face
3. two neighbouring tetrahedra $P_{\alpha}$ and $P_{\alpha''}$ sharing an edge
4. two neighbouring tetrahedra $P_{\alpha}$ and $P_{\alpha'''}$ sharing a point 
5. two distinct tetrahedra $P_{\alpha}$ and $P_{\beta}$
6. reduce 6D integrals to sum of 4D integrals (how many exactly?) using Euler's theorem for homogeneous functions;
7. evaluate the 4D integrals numerically by some adaptive quadrature;    

## Section 2: Edges, EdgePairs, Facets, PointFacet Pair and Linear Functions 

In [4]:
Base.@kwdef struct Edge3D 
    r1::Point3D
    r2::Point3D
    len::Float64 = norm(r2-r1)
    tau::Point3D = normalize(r2-r1)
    e::Point3D = tau/len 
end 

function (edge3D::Edge3D)(r)
    return [dot(edge3D.r2-r,edge3D.e), dot(r-edge3D.r1,edge3D.e)] 
end

# here we assume lines to intersect 
function edge3Dintersection(el1::Edge3D,el2::Edge3D)
    rdiff = el2.r1-el1.r1
    num1 = dot(el1.tau,el2.tau)*dot(rdiff,el2.tau)-dot(rdiff,el1.tau)
    num2 = -dot(el1.tau,el2.tau)*dot(rdiff,el1.tau)+dot(rdiff,el2.tau)
    den  = dot(el1.tau,el2.tau)^2-1
    return [el1.r1 + (num1/den)*el1.tau, el2.r1 + (num2/den)*el2.tau]
end

# super-type to hold a pair of edges 
abstract type Edge3DPair end 

# intersecting pair of edges 
Base.@kwdef struct InterEdge3DPair <: Edge3DPair
    e1::Edge3D                                # first edge 
    e2::Edge3D                                # second edge 
    s::Point3D = edge3Dintersection(e1,e2)[1] # point of intersection 
end 

# crossing pair of edges 
Base.@kwdef struct CrossEdge3DPair <: Edge3DPair
    e1::Edge3D                                     # first edge 
    e2::Edge3D                                     # second edge 
    nw::Point3D = normalize(cross(e1.tau,e2.tau))  # normal on the plane
    h::Float64 = abs(dot(nw,e1.r1-e2.r1))          # height between the edges 
    r1hat::Point3D = e1.r1+dot(e2.r1-e1.r1,nw)*nw  # projection of start of first edge onto plane 
    r2hat::Point3D = e1.r2+dot(e2.r2-e1.r2,nw)*nw  # projection of end of first edge onto plane
    e1hat::Edge3D = Edge3D(r1 = r1hat,r2 = r2hat)  # projected first edge 
    s::Point3D = edge3Dintersection(e1hat,e2)[1]   # point of intersection 
end

# parallel pair of edges
struct ParalEdge3DPair <: Edge3DPair end

# inline pair of edges    
struct InlinEdge3DPair <: Edge3DPair end

Base.@kwdef struct Facet3D 
    r1::Point3D                                # first vertex of the facet  
    r2::Point3D                                # second vertex of the facet
    r3::Point3D                                # third vertex of the facet
    tau1::Point3D = normalize(r2 - r1)         # direction of first edge   
    tau2::Point3D = normalize(r3 - r2)         # direction of second edge
    tau3::Point3D = normalize(r1 - r3)         # direction of third edge
    area::Float64 = .5*norm(cross(tau1,tau2))  # area of triangle 
    nF::Point3D = normalize(cross(tau1,tau2))  # normal to face 
    n1::Point3D = normalize(cross(tau1,nF))    # normal on first edge   
    n2::Point3D = normalize(cross(tau2,nF))    # normal on second edge
    n3::Point3D = normalize(cross(tau3,nF))    # normal on third edge
end 

Base.@kwdef struct PointFacet3DPair 
    a::Point3D                               
    F::Facet3D
    ahat::Point3D = a + dot(F.r1-a,F.nF)*F.nF
    h::Float64 = abs(dot(F.r1-a,F.nF))
    smin1::Float64 = dot(F.r1-ahat,F.tau1)
    smin2::Float64 = dot(F.r2-ahat,F.tau2)
    smin3::Float64 = dot(F.r3-ahat,F.tau3)
    smax1::Float64 = dot(F.r2-ahat,F.tau1)
    smax2::Float64 = dot(F.r3-ahat,F.tau2)
    smax3::Float64 = dot(F.r1-ahat,F.tau3)
    R01::Float64   = norm(cross(ahat-F.r1,F.tau1)) 
    R02::Float64   = norm(cross(ahat-F.r2,F.tau2))
    R03::Float64   = norm(cross(ahat-F.r3,F.tau3))
end 

Base.@kwdef struct FacetFacet3DPair 
    F1::Facet3D
    F2::Facet3D
end 

struct LinearFunction  
    a::Float64
    b::Float64    
    c::Float64
    d::Float64
end 

function (linearFunction::LinearFunction)(r)
    a = linearFunction.a
    b = linearFunction.b
    c = linearFunction.c
    d = linearFunction.d
    return [a*r[1]+linearFunction.b*r[2]+c*r[3]+d, [a,b,c]] 
end

In [5]:
# define an edge 
r1 = Point3D(1.,0.,0.,); r2 = Point3D(2.,0.,0.); 
el1 = Edge3D(r1=r1,r2=r2)

Edge3D([1.0, 0.0, 0.0], [2.0, 0.0, 0.0], 1.0, [1.0, 0.0, 0.0], [1.0, 0.0, 0.0])

In [6]:
# returns value of two basis functions associated with the edge   
s = Point3D(2.,0.,0.)
el1(s)

2-element Vector{Float64}:
 0.0
 1.0

#### Data for intersecting edges 

In [7]:
# here we assume lines to intersect 
function edge3Dintersection(el1::Edge3D,el2::Edge3D)
    rdiff = el2.r1-el1.r1
    num1 = dot(el1.tau,el2.tau)*dot(rdiff,el2.tau)-dot(rdiff,el1.tau)
    num2 = -dot(el1.tau,el2.tau)*dot(rdiff,el1.tau)+dot(rdiff,el2.tau)
    den  = dot(el1.tau,el2.tau)^2-1
    return [el1.r1 + (num1/den)*el1.tau, el2.r1 + (num2/den)*el2.tau]
end

# intersecting pair of edges 
Base.@kwdef struct InterEdge3DPair6 <: Edge3DPair 
    e1::Edge3D
    e2::Edge3D
    s::Point3D = edge3Dintersection(e1,e2)[1] # point of intersection 
end 

InterEdge3DPair6

In [8]:
# define first edge 
r11 = Point3D(1.,1.,0.,); r12 = Point3D(2.,2.,0.); 
#r11 = Point3D(0.,1.,0.,); r12 = Point3D(0.,2.,0.); 
el1 = Edge3D(r1 = r11,r2 = r12)

# define second edge 
r21 = Point3D(1.,0.,0.,); r22 = Point3D(2.,0.,0.);
el2 = Edge3D(r1 = r21,r2 = r22)

interEdge3DPair = InterEdge3DPair6(e1 = el1, e2 = el2)

InterEdge3DPair6(Edge3D([1.0, 1.0, 0.0], [2.0, 2.0, 0.0], 1.4142135623730951, [0.7071067811865475, 0.7071067811865475, 0.0], [0.49999999999999994, 0.49999999999999994, 0.0]), Edge3D([1.0, 0.0, 0.0], [2.0, 0.0, 0.0], 1.0, [1.0, 0.0, 0.0], [1.0, 0.0, 0.0]), [3.3306690738754696e-16, 3.3306690738754696e-16, 0.0])

#### Data for crossing edges  

In [9]:
# define first edege 
r11 = Point3D(1.,1.,1.,); r12 = Point3D(2.,2.,1.); 
el1 = Edge3D(r1 = r11,r2 = r12)

# define second edge 
r21 = Point3D(1.,0.,0.,); r22 = Point3D(2.,0.,0.); 
el2 = Edge3D(r1 = r21,r2 = r22)

crossEdge3DPair = CrossEdge3DPair(e1 = el1, e2 = el2)

CrossEdge3DPair(Edge3D([1.0, 1.0, 1.0], [2.0, 2.0, 1.0], 1.4142135623730951, [0.7071067811865475, 0.7071067811865475, 0.0], [0.49999999999999994, 0.49999999999999994, 0.0]), Edge3D([1.0, 0.0, 0.0], [2.0, 0.0, 0.0], 1.0, [1.0, 0.0, 0.0], [1.0, 0.0, 0.0]), [0.0, 0.0, -1.0], 1.0, [1.0, 1.0, 0.0], [2.0, 2.0, 0.0], Edge3D([1.0, 1.0, 0.0], [2.0, 2.0, 0.0], 1.4142135623730951, [0.7071067811865475, 0.7071067811865475, 0.0], [0.49999999999999994, 0.49999999999999994, 0.0]), [3.3306690738754696e-16, 3.3306690738754696e-16, 0.0])

#### Data for a Facet 

In [10]:
r1 = Point3D(0.,0.,0.); r2 = Point3D(1.,0.,0.); r3 = Point3D(0.,1.,0.);
facet = Facet3D(r1=r1, r2=r2, r3=r3) 

Facet3D([0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [1.0, 0.0, 0.0], [-0.7071067811865475, 0.7071067811865475, 0.0], [0.0, -1.0, 0.0], 0.35355339059327373, [0.0, -0.0, 1.0], [0.0, -1.0, -0.0], [0.7071067811865476, 0.7071067811865476, 0.0], [-1.0, 0.0, 0.0])

In [11]:
facet.tau1

3-element SVector{3, Float64} with indices SOneTo(3):
 1.0
 0.0
 0.0

#### Data for PointFacet3D Pair 

In [12]:
a = Point3D(0.,0.,10.);
r1 = Point3D(0.,0.,0.); r2 = Point3D(1.,0.,0.); r3 = Point3D(0.,1.,0.);
facet = Facet3D(r1=r1, r2=r2, r3=r3)
pointfacetpair = PointFacet3DPair(a = a, F = facet) 

PointFacet3DPair([0.0, 0.0, 10.0], Facet3D([0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [1.0, 0.0, 0.0], [-0.7071067811865475, 0.7071067811865475, 0.0], [0.0, -1.0, 0.0], 0.35355339059327373, [0.0, -0.0, 1.0], [0.0, -1.0, -0.0], [0.7071067811865476, 0.7071067811865476, 0.0], [-1.0, 0.0, 0.0]), [0.0, 0.0, 0.0], 10.0, 0.0, -0.7071067811865475, -1.0, 1.0, 0.7071067811865475, 0.0, 0.0, 0.7071067811865475, 0.0)

In [13]:
#[pointfacetpair.R01, pointfacetpair.R02, pointfacetpair.R03] 
pointfacetpair.smin2

-0.7071067811865475

#### Linear Function Testing 

In [14]:
phi = LinearFunction(1.,2.,3.,4.)
r = Point3D(1.,-1.,1.,)
display(phi(r)[1]) 
display(phi(r)[2])

6.0

3-element Vector{Float64}:
 1.0
 2.0
 3.0

## Section 3: Integrals 

### Section 3.1: Point-Point Interactions (implement f4 and f5)
### Implement I14hat, I15hat, I16hat and I17hat for parallel facets (odd powers easier to implement)  

In [74]:
function f1(h,eta)
    return 1/eta^2*(sqrt(h^2+eta^2)-h)
end 

function f2(h,eta)
    return sqrt(h^2+eta^2)/(2*eta^2)-h^2/(2*eta^3)*asinh(eta/h)
end 

function f3(h,eta)
    return (2*h^3+(h^2+eta^2)^(3/2)-3*h^2*sqrt(h^2+eta^2))/(3*eta^4)
end

function f4(h,eta)
    return ((2*eta^3-3*eta*h^2)*sqrt(h^2+eta^2)+3*h^4*asinh(eta/h))/(8*eta^5)
end 

function f5(h,eta)
    return (3*(h^2+eta^2)^(2.5)-10*h^2*(h^2+eta^2)^(1.5)+15*h^4*sqrt(h^2+eta^2)-8*h^5)/(15*eta^6)
end 

function I8integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return f1(h,arg) 
end

function I8(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return I14(s,h,R0) + asinh(s/t0) - h/R0*atan(s/R0)
end

function I9(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return sqrt(t0^2+s^2)/h-asinh(h/sqrt(R0^2+s^2))-.5*log(R0^2+s^2) 
end

function I9integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return s/h*f1(h,arg) 
end

function I10(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return t0^2/(2*R0^2)*asinh(s/t0)-(h^2*s)/(2*R0^2*sqrt(R0^2+s^2))*asinh(sqrt(R0^2+s^2)/h)
end

function I10integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return f2(h,arg) 
end

function I11(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return sqrt(t0^2+s^2)/(2*h)+h*asinh(sqrt(R0^2+s^2)/h)/(2*sqrt(R0^2+s^2))
end

function I11integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return s/h*f2(h,arg) 
end

function I12(s,h,R0) 
    return I14(s,h,R0)/3+I15(s,h,R0)/3+I18(s,h,R0)+2*I19(s,h,R0)/3 
end

function I12integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return f3(h,arg) 
end

function I13(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return ((t0^2+s^2)^(3/2)/h-h^2)/(3*(R0^2+s^2))
end

function I13integrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return s/h*f3(h,arg) 
end

function I14(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    argplus = sqrt(t0^2+s^2)+h
    argmin  = sqrt(t0^2+s^2)-h
    return sign(s)*(I17plus(argmin,h,R0) - I17min(argplus,h,R0)) 
end

function I15(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return asinh(s/t0)
end

function I16(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return h/R0*atan(s/R0)
end

function I17plus(x,h,R0)
    t0 = sqrt(h^2+R0^2)
    arg = -R0^2/(x*t0)+h/t0
    return h/(2*R0)*asin(arg) 
end 

function I17min(x,h,R0)
    t0 = sqrt(h^2+R0^2)
    arg  = -R0^2/(x*t0)-h/t0
    return h/(2*R0)*asin(arg)  
end

function I18(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    return h^3/(3*R0^3)*(R0*s/(R0^2+s^2)+atan(s/R0)) 
end

function I19(s,h,R0)
    t0 = sqrt(h^2+R0^2)
    value = -(h^2*s*sqrt(t0^2+s^2))/(2*R0^2*(R0^2+s^2))
    value += h/(2*R0)*(1-h^2/R0^2)*atan(h*s/(R0*sqrt(t0^2+s^2)))
    return value-I14(s,h,R0)
end

function I14hat(s,h,R0) 
    t0 = sqrt(h^2+R0^2)
    t1 = 1/8*(1+h^2/R0^2)*I14(s,h,R0)
    t2 = 1/4*(1-h^4/R0^4)*I15(s,h,R0)
    t3 = 1/4*I19(s,h,R0)
    t4 = h^4/(8*R0^4)*(R0^2*s/(R0^2+s^2)^(1.5)+2*s/sqrt(R0^2+s^2))*asinh(sqrt(R0^2+s^2)/h)
    return t1+t2+t3+t4 
end

function I14hatintegrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return f4(h,arg) 
end

function I15hat(s,h,R0) 
    t0 = sqrt(h^2+R0^2)
    return sqrt(t0^2+s^2)/(4*h)+h*sqrt(t0^2+s^2)/(8*(R0^2+s^2))-h^3*asinh(sqrt(R0^2+s^2)/h)/(8*(R0^2+s^2)^(1.5))
end

function I15hatintegrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return s/h*f4(h,arg) 
end

function I16hat(s,h,R0) 
    t0 = sqrt(h^2+R0^2)
    t1 = 2*h^2/(15*R0^2)*s*(t0^2+s^2)^(1.5)/(R0^2+s^2)^2 
    t2 = 1/5*asinh(s/t0)
    t3 = (h^4/(5*R0^4)-h^2/(5*R0^2))*s*sqrt(t0^2+s^2)/(R0^2+s^2)
    t4 = (h^5/(5*R0^5)+h/(15*R0))*atan(h*s/(R0*sqrt(t0^2+s^2)))
    t5 = -1/15*I14(s,h,R0)
    t6 = -h^5/(15*R0^5)*(2*R0^3*s/(R0^2+s^2)^2+3*R0*s/(R0^2+s^2)+3*atan(s/R0))
    return t1+t2+t3+t4+t5+t6 
end

function I16hatintegrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return f5(h,arg) 
end

function I17hat(s,h,R0) 
    t0 = sqrt(h^2+R0^2)
    return (2*h)/15*(h^3-(t0^2+s^2)^(1.5))/(R0^2+s^2)^2+h/5*sqrt(t0^2+s^2)/(R0^2+s^2)+1/(5*h)*sqrt(t0^2+s^2)
end

function I17hatintegrand(s,h,R0)
    arg = sqrt(R0^2+s^2)
    return s/h*f5(h,arg) 
end

I17hatintegrand (generic function with 1 method)

### Section 3.2: Point-Edge Interactions

In [16]:
function I3(a::Point3D,el::Edge3D)
    num  = norm(el.r2-a) + norm(el.r1-a) + norm(el.r2-el.r1)
    den  = norm(el.r2-a) + norm(el.r1-a) - norm(el.r2-el.r1)
    return log(num/den)  
end 

function I3integrand(lmbd,a::Point3D,el1::Edge3D)
    r  = el1.r1+lmbd*el1.tau   
    return 1/norm(r-a)
end

function Rphi1(a::Point3D,el::Edge3D)
    t1 = dot(el.r2-a,el.tau)/norm(el.r2-el.r1)*I3(a,el)
    t2 = (norm(el.r1-a) - norm(el.r2-a)) /norm(el.r2-el.r1)
    return t1+t2 
end 

function Rphi1integrand(lmbd,a::Point3D,el::Edge3D)
    r  = el1.r1+lmbd*el1.tau
    return dot(el.r2-r,el.e)/norm(r-a)
end;

In [17]:
# define a point
a = Point3D(0.5,1e-1,0.,);

# define an edge 
r11 = Point3D(0.,0.,0.,); r12 = Point3D(1.,0.,0.);
el1 = Edge3D(r1 = r11,r2 = r12) 

# test the computation of I3
I3val = I3(a,el1) 
I3valnum = hcubature(lmbd -> I3integrand(lmbd[1],a,el1), (0,), (el1.len,);rtol=1e-6,atol=1e-6)[1]
@test I3val ≈ I3valnum  

# test the computation of Rphi1 
Rphi1val = Rphi1(a,el1) 
Rphi1valnum = hcubature(lmbd -> Rphi1integrand(lmbd[1],a,el1), (0,), (el1.len,);rtol=1e-6,atol=1e-6)[1]
@test Rphi1val ≈ Rphi1valnum

Test Passed

### Section 3.3: Edge-Edge Interactions

Here we implement the line - line interaction 

$$ 
Q^{\phi_1,\theta_1}(l_1,l_2) = \int_{l_1} \int_{l_2} \frac{\phi_1(\mathbf{r}) \, \theta_1(\mathbf{r}')}
                                                         {\| \mathbf{r} - \mathbf{r}' \|}dl_1 \, dl_2' 
$$

This integral is singular in case that $l_1$ and $l_2$ intersect on the domain of integration. In this case, numerical integration might become cumbersome. We distinguish the following 4 cases 

1. case-1: intersecting lines: see below;
2. case-2: crossing lines: $Q^{\phi_1,\theta_1} = I_{31} + I_{32} + I_{33} + I_{34}$, where $I_{34}$ can be computed in terms of $I_8$ (see earlier), where $I_{31}$ is defined in terms of $I_{35}$, $I_{36}$, $I_{37}$ and $I_{38}$ and where $I_{32}$ and $I_{33}$ afre similar to each other; 
3. case-3: parallel lines: to be described; 
4. case-4: coinciding lines or as extension of each other: to be described;

#### Case-1: Intersecting Lines 

In [18]:
function I25(s::Point3D,el1::Edge3D,el2::Edge3D)
    t1 = dot(el2.r2-s,el2.tau)/el2.len*I3(el1.r2,el2) - Rphi1(el1.r2,el2)
    t1 = t1*norm(el1.r2-s)^2/norm(el1.r2-el1.r1)
    t2 = dot(el2.r2-s,el2.tau)/el2.len*I3(el1.r1,el2) - Rphi1(el1.r1,el2)
    t2 = t2*norm(el1.r1-s)^2/norm(el1.r2-el1.r1)
    t3 = dot(el1.r2-s,el1.tau)/el1.len*I3(el2.r2,el1) - Rphi1(el2.r2,el1)
    t3 = t3*norm(el2.r2-s)^2/norm(el2.r2-el2.r1)
    t4 = dot(el1.r2-s,el1.tau)/el1.len*I3(el2.r1,el1) - Rphi1(el2.r1,el1)
    t4 = t4*norm(el2.r1-s)^2/norm(el2.r2-el2.r1)
    return (t1-t2+t3-t4)/3 
end

function I25integrand(lmbd1,lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D)
    r  = el1.r1+lmbd1*el1.tau 
    rp = el2.r1+lmbd2*el2.tau
    # return (dot(s-r,el1.e)+el1(s)[1])*(dot(s-rp,el2.e)+el2(s)[1])/norm(r-rp)
    return (dot(s-r,el1.e))*(dot(s-rp,el2.e))/norm(r-rp)
end

function I26(s::Point3D,el1::Edge3D,el2::Edge3D)
    t1 = norm(el2.r2-s)^2/(2*el2.len)*I3(el2.r2,el1)
    t2 = norm(el2.r1-s)^2/(2*el2.len)*I3(el2.r1,el1)
    t3 = dot(el1.r2-s,el1.tau)/2*(dot(el2.r2-s,el2.tau)/el2.len*I3(el1.r2,el2)-Rphi1(el1.r2,el2))
    t4 = dot(el1.r1-s,el1.tau)/2*(dot(el2.r2-s,el2.tau)/el2.len*I3(el1.r1,el2)-Rphi1(el1.r1,el2)) 
    return t1-t2+t3-t4
end

function I26integrand(lmbd1,lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D)
    r  = el1.r1+lmbd1*el1.tau 
    rp = el2.r1+lmbd2*el2.tau
    return dot(rp-s,el2.e)/norm(r-rp)
end

function I27(s::Point3D,el1::Edge3D,el2::Edge3D)
    t1 = norm(el1.r2-s)^2/(2*el1.len)*I3(el1.r2,el2)
    t2 = norm(el1.r1-s)^2/(2*el1.len)*I3(el1.r1,el2)
    t3 = dot(el2.r2-s,el2.tau)/2*(dot(el1.r2-s,el1.tau)/el1.len*I3(el2.r2,el1)-Rphi1(el2.r2,el1))
    t4 = dot(el2.r1-s,el2.tau)/2*(dot(el1.r2-s,el1.tau)/el1.len*I3(el2.r1,el1)-Rphi1(el2.r1,el1))
    return t1-t2+t3-t4 
end

function I27integrand(lmbd1,lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D)
    r  = el1.r1+lmbd1*el1.tau 
    rp = el2.r1+lmbd2*el2.tau
    return dot(r-s,el1.e)/norm(r-rp)
end

function I28(s::Point3D,el1::Edge3D,el2::Edge3D)
    t1 = dot(el1.r2-s,el1.tau)*I3(el1.r2,el2)
    t2 = dot(el1.r1-s,el1.tau)*I3(el1.r1,el2)
    t3 = dot(el2.r2-s,el2.tau)*I3(el2.r2,el1)
    t4 = dot(el2.r1-s,el2.tau)*I3(el2.r1,el1)
    return t1-t2+t3-t4
end

function I28integrand(lmbd1,lmbd2,el1::Edge3D,el2::Edge3D)
    r  = el1.r1+lmbd1*el1.tau 
    rp = el2.r1+lmbd2*el2.tau
    return 1/norm(r-rp)
end

function Qphi1theta1(el1::Edge3D,el2::Edge3D)
    s = edge3Dintersection(el1,el2)[1]
    #return I25(s,el1,el2)-el1(s)[1]*I26(s,el1,el2)-el2(s)[1]*I27(s,el1,el2)+el1(s)[1]*el2(s)[1]*I28(s,el1,el2)
    return I25(s,el1,el2)-el1(s)[1]*I26(s,el1,el2)-el2(s)[1]*I27(s,el1,el2)+el1(s)[1]*el2(s)[1]*I28(s,el1,el2)
end

function Qphi1theta1integrand(lmbd1,lmbd2,el1::Edge3D,el2::Edge3D)
    r  = el1.r1+lmbd1*el1.tau 
    rp = el2.r1+lmbd2*el1.tau
    # display( (el1(r)[1])*(el2(rp)[1])/norm(r-rp) )   
    return (el1(r)[1])*(el2(rp)[1])/norm(r-rp)
end;

### Testing of Intersecting Edge - Edge Interaction 

In [19]:
# define a first edge
r11 = Point3D(1.,1.,0.,); r12 = Point3D(2.,2.,0.); len = norm(r12-r11)
el1 = Edge3D(r1 = r11, r2 = r12) 

# define a second edge
r21 = Point3D(1.,0.,0.,); r22 = Point3D(2.,0.,0.); len = norm(r22-r21)
el2 = Edge3D(r1 = r21, r2 = r22)

# compute the intersection of the two lines 
out = edge3Dintersection(el1,el2);

In [20]:
I25val = I25(out[1],el1,el2)
I26val = I26(out[1],el1,el2)
I27val = I27(out[1],el1,el2)
I28val = I28(out[1],el1,el2)

I25valnum = hcubature(lmbd -> I25integrand(lmbd[1],lmbd[2],out[1],el1,el2), (0,0), (el1.len,el2.len);rtol=1e-10,atol=1e-10)[1]  
I26valnum = hcubature(lmbd -> I26integrand(lmbd[1],lmbd[2],out[1],el1,el2), (0,0), (el1.len,el2.len);rtol=1e-10,atol=1e-10)[1]  
I27valnum = hcubature(lmbd -> I27integrand(lmbd[1],lmbd[2],out[1],el1,el2), (0,0), (el1.len,el2.len);rtol=1e-10,atol=1e-10)[1]  
I28valnum = hcubature(lmbd -> I28integrand(lmbd[1],lmbd[2],el1,el2), (0,0), (el1.len,el2.len);rtol=1e-10,atol=1e-10)[1]  

@test I25val ≈ I25valnum 
@test I26val ≈ I26valnum
@test I27val ≈ I27valnum 
@test I28val ≈ I28valnum

Test Passed

In [21]:
Qphi1theta1(el1,el2)

0.26506692155603107

In [22]:
hcubature(s -> Qphi1theta1integrand(s[1],s[2],el1,el2), (0,0), (el1.len,el2.len);rtol=1e-6,atol=1e-6)  

(0.44924713390553317, 9.723066444582151e-7)

#### Case-2: Crossing Lines 

In [23]:
function I31()
    return 0. 
end

function I31integrand(lmbd1,lmbd2,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
    rp  = el2.r1+lmbd2*el2.tau
    num = dot(r,el1.e)*dot(rp,el2.e)
    den = sqrt(h*2+norm(r-rp)^2)
    return num/den
end

function I32()
    return 0. 
end

function I32integrand(lmbd1,lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
    rp  = el2.r1+lmbd2*el2.tau
    num = dot(r-s,el1.e)
    den = sqrt(h*2+norm(r-rp)^2)
    return num/den
end

function I33()
    return 0. 
end

function I33integrand(lmbd1,lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
    rp  = el2.r1+lmbd2*el2.tau
    num = dot(rp-s,el2.e)
    den = sqrt(h*2+norm(r-rp)^2)
    return num/den    
end

function I34()
    return 0. 
end

function I34integrand(lmbd1,lmbd2,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
    rp  = el2.r1+lmbd2*el2.tau
    return 1/sqrt(h*2+norm(r-rp)^2)
end

function I35()
    return 0. 
end

function I35integrand(lmbd2,s::Point3D,el1::Edge3D,el2::Edge3D,h)
    rp  = el2.r1+lmbd2*el2.tau
    return dot(rp-s,el2.e)*f3(h,norm())
end

function I36()
end

function I36integrand(lmbd2,el1::Edge3D,el2::Edge3D,h)
    rp  = el2.r1+lmbd2*el2.tau
    return 0.
end

function I37()
    return 0. 
end

function I37integrand(lmbd1,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
end

function I38()
    return 0. 
end

function I38integrand(lmbd1,el1::Edge3D,el2::Edge3D,h)
    r   = el1.r1+lmbd1*el1.tau 
end

I38integrand (generic function with 1 method)

#### Case-3: Parallel Lines

In [24]:
function I200()
    return 0. 
end

function I200integrand()
    return 0.
end

I200integrand (generic function with 1 method)

#### Case-3: Coinciding Lines or as Extension of Each Other 
* Analytically: currently fails as I3() returns Inf for point belonging to line;
* Numerically: currently fails as integrand singular neaer the lower bound; 

In [25]:
function I41(el1::Edge3D,el2::Edge3D)
    return el1(el1.r2)[1]*I3(el1.r1,el2) - Rphi1(el1.r1,el2)
end

function I41integrand(lmbd2,el1::Edge3D,el2::Edge3D)
    rp  = el2.r1+lmbd2*el2.tau
    return (el1(el1.r2)[1]-el1(rp)[1])/norm(el1.r1-rp)
end

I41integrand (generic function with 1 method)

In [26]:
# define first edege 
r11 = Point3D(1.,1.,1.,); r12 = Point3D(2.,2.,1.); 
el1 = Edge3D(r1 = r11, r2 = r12)

el2 = deepcopy(el1)

Edge3D([1.0, 1.0, 1.0], [2.0, 2.0, 1.0], 1.4142135623730951, [0.7071067811865475, 0.7071067811865475, 0.0], [0.49999999999999994, 0.49999999999999994, 0.0])

In [27]:
I41(el1,el2)

NaN

In [28]:
el1(el1.r2)[1]

0.0

In [29]:
I3(el1.r1,el2)

Inf

In [30]:
Rphi1(el1.r1,el2)

Inf

In [31]:
I41integrand(1.,el1,el2)

-0.29289321881345254

In [32]:
hcubature(s -> I41integrand(s[1],el1,el2), (0.1,), (el1.len,);rtol=1e-6,atol=1e-6)  

(-1.7198693613937854, 8.201437841914849e-7)

### Section 3.3: Point - Facet Interactions

Here we implement the point - facet interaction 

$$
P^{\phi}({\mathbf a}, F) = \iint_{F} \frac{\phi({\mathbf a})}{\| {\mathbf r} - {\mathbf a}\|} \, dS \, . 
$$
 
In case of triangular facets, $P^{\phi}$ is computed as a sum of 3 contributions, one for each edge of the triangle. This integral is singular in case that ${\mathbf a}$ lies on F. A seperate analytical solution for this case is available.  

In [33]:
function P1(pointfacet3Dpair::PointFacet3DPair,linearfunction::LinearFunction)

    ahat  = pointfacet3Dpair.ahat
    h     = pointfacet3Dpair.h 
    F     = pointfacet3Dpair.F 
    smin1 = pointfacet3Dpair.smin1
    smin2 = pointfacet3Dpair.smin2 
    smin3 = pointfacet3Dpair.smin3 
    smax1 = pointfacet3Dpair.smax1
    smax2 = pointfacet3Dpair.smax2 
    smax3 = pointfacet3Dpair.smax3 
    R01   = pointfacet3Dpair.R01 
    R02   = pointfacet3Dpair.R02
    R03   = pointfacet3Dpair.R03 

    phival  = linearfunction(ahat)[1]    
    phigrad = linearfunction(ahat)[2]

    t1 = 0.   
    if !(R01==0.) 
        t1  = dot(F.r1-smin1*F.tau1-ahat,phigrad)*I10(smax1,h,R01)
        t1 += h*dot(F.tau1,phigrad)*I11(smax1,h,R01)+phival*I8(smax1,h,R01) 
        t1 -= dot(F.r1-smin1*F.tau1-ahat,phigrad)*I10(smin1,h,R01)
        t1 -= h*dot(F.tau1,phigrad)*I11(smin1,h,R01)+phival*I8(smin1,h,R01)
        t1 *= dot(F.r1-ahat,F.n1)
    end 

    t2 = 0.   
    if !(R02==0.) 
        t2  = dot(F.r2-smin2*F.tau2-ahat,phigrad)*I10(smax2,h,R02)
        t2 += h*dot(F.tau2,phigrad)*I11(smax2,h,R02)+phival*I8(smax2,h,R02)  
        t2 -= dot(F.r2-smin2*F.tau2-ahat,phigrad)*I10(smin2,h,R02)
        t2 -= h*dot(F.tau2,phigrad)*I11(smin2,h,R02)+phival*I8(smin2,h,R02)     
        t2 *= dot(F.r2-ahat,F.n2)
    end 

    t3 = 0.
    if !(R02==0.)
        t3  = dot(F.r3-smin2*F.tau3-ahat,phigrad)*I10(smax3,h,R03)
        t3 += h*dot(F.tau3,phigrad)*I11(smax3,h,R03)+phival*I8(smax3,h,R03)  
        t3 -= dot(F.r3-smin3*F.tau3-ahat,phigrad)*I10(smin3,h,R03)
        t3 -= h*dot(F.tau3,phigrad)*I11(smin3,h,R03)+phival*I8(smin3,h,R03)     
        t3 *= dot(F.r3-ahat,F.n3)
    end 
    
    return t1+t2+t3
end

function P1phiconst(pointfacet3Dpair::PointFacet3DPair,linearfunction::LinearFunction)

    ahat  = pointfacet3Dpair.ahat
    h     = pointfacet3Dpair.h 
    F     = pointfacet3Dpair.F 
    smin1 = pointfacet3Dpair.smin1
    smin2 = pointfacet3Dpair.smin2 
    smin3 = pointfacet3Dpair.smin3 
    smax1 = pointfacet3Dpair.smax1
    smax2 = pointfacet3Dpair.smax2 
    smax3 = pointfacet3Dpair.smax3 
    R01   = pointfacet3Dpair.R01 
    R02   = pointfacet3Dpair.R02
    R03   = pointfacet3Dpair.R03 

    t1 = 0.; 
    if !(R01==0.) 
        t1  = I8(smax1,h,R01)-I8(smin1,h,R01)  
        t1 *= dot(F.r1-ahat,F.n1)
    end 

    t2 = 0.;
    if !(R02==0.)
       t2  = I8(smax2,h,R02)-I8(smin2,h,R02)         
       t2 *= dot(F.r2-ahat,F.n2)
    end 

    t3 = 0.;
    if !(R03==0.)
        t3  = I8(smax3,h,R03)-I8(smin3,h,R03)
        t3 *= dot(F.r3-ahat,F.n3)
    end 
    
    return t1+t2+t3
end

function P1integrand(pointfacetpair::PointFacet3DPair)
    a = pointfacetpair.a
    F = pointfacetpair.F  
    return 0.
end

P1integrand (generic function with 1 method)

In [34]:
# test special case of constant function 
a = Point3D(0.25,.25,.25);
r1 = Point3D(0.,0.,0.); r2 = Point3D(1.,0.,0.); r3 = Point3D(0.,1.,0.);
facet = Facet3D(r1=r1, r2=r2, r3=r3)
pointfacetpair = PointFacet3DPair(a = a, F = facet);
phi = LinearFunction(0.,0.,0.,1.)
@test P1(pointfacetpair,phi) ≈ P1phiconst(pointfacetpair,phi)

Test Passed

In [35]:
a = Point3D(0.,0.,1.);
r1 = Point3D(0.,0.,0.); r2 = Point3D(1.,0.,0.); r3 = Point3D(0.,1.,0.);
facet = Facet3D(r1=r1, r2=r2, r3=r3)
pointfacetpair = PointFacet3DPair(a = a, F = facet)
phi = LinearFunction(0.,0.,0.,1.)
P1phiconst(pointfacetpair,phi)
# P1(pointfacetpair,phi)

0.4369992897579717

In [68]:
pointfacetpair.F.tau3

3-element SVector{3, Float64} with indices SOneTo(3):
  0.0
 -1.0
  0.0

In [64]:
pointfacetpair.R02

0.7071067811865475

In [56]:
pointfacetpair.smax2

0.7071067811865475

In [52]:
pointfacetpair.F.n2

3-element SVector{3, Float64} with indices SOneTo(3):
 0.7071067811865476
 0.7071067811865476
 0.0

In [70]:
I8(pointfacetpair.smin2,pointfacetpair.h,pointfacetpair.R02)

-0.30900516116156673

In [71]:
sqrt(2)/2*(I8(pointfacetpair.smax2,pointfacetpair.h,pointfacetpair.R02) - I8(pointfacetpair.smin2, pointfacetpair.h,pointfacetpair.R02))

0.4369992897579717

In [73]:
a

3-element SVector{3, Float64} with indices SOneTo(3):
 0.0
 0.0
 1.0

In [72]:
# attempt to compute Pphi numerically  
hcubature(x -> x[1]/sqrt((x[1]-a[1])^2+(x[1]*x[2]-a[2])^2+a[3]^2), (0,0), (1,1))

(0.39667956065749077, 5.569539877903207e-9)

In [77]:
# does evaluating I8 for h=0 pose annty challenges? 
I8(1,0,1)

0.881373587019543

## Small Test Case for point-facet interaction $P^1(a,F)$

Assume triangular facet in the $xy$-plane with nodes $\vec{r}_1 = (0,0,0)$, $\vec{r}_2 = (1,0,0)$ and $\vec{r}_3 = (0,1,0)$. Then $\vec{n}_1 = (0,-1,0)$, $\vec{n}_2 = (\sqrt{2}/2,\sqrt{2}/2,0)$ and $\vec{n}_3 = (-1,0,0)$ and $R_0^1 = 0$, $R_0^2 = \sqrt{2}/2 $ and $R_0^3 = 0$. Then only the second edge contributes to the integral. Over the second edge, we have that $-\sqrt{2}/2\leq s \leq \sqrt{2}/2$. 

Assume point $\vec{a}$ on the $z$-axis with coordinates $\vec{a} = (0,0,1)$. Then $\vec{\widehat{a}} = (0,0,0)$, $h=1$ and $\|  \vec{a} - \vec{r}\| = \sqrt{x^2+y^2+1}$. 

Assume the function $\varphi(\vec{x})$ to be constant equal one, i.e., $\varphi(\vec{x}) = 1$.  

Numerical integration using hcubature.jl gives 

$$ 
P^1(a,F) = \iint_{F} \frac{1}{\| \vec{a} - \vec{r}\|} \, dS = 
\int_0^1 \left[ \int_0^{1-x} \frac{1}{\sqrt{x^2+y^2+1}} dy \right] dx = 
- \int_0^1  \text{asinh} \left[ \frac{x}{\sqrt{x^2+1}} - \frac{1}{\sqrt{x^2+1}}\right] \, dx = 0.437\ldots \, . 
$$

Symbolic computation instead gives (only the second edge contributes)

$$ 
P^1(a,F) = \iint_{F} \frac{1}{\| \vec{a} - \vec{r}\|} \, dS =
(\vec{r}_2 - \vec{\widehat{a}}) \cdot \vec{n}_2 \left[ I_8(s,h=1,R_0^2=\sqrt{2}/2 \right]_{s=-\sqrt{2}/2}^{s=\sqrt{2}/2} = 
\sqrt{2} / 2 \left[ I_8(s,h=1,R_0^2=\sqrt{2}/2 \right]_{s=-\sqrt{2}/2}^{s=\sqrt{2}/2} = 0.437\ldots \, . 
$$

Numerical and analytical answers do match. 

## Section 4: Testing Function f1 through f5  

In [58]:
f5(.5,.5)

0.25326599154822227

## Section 5: Testing Integrals I8 up to I17hat 

In [75]:
h = .5
R0 = .5
s1 = .1; s2 = 1. 
I8val      = I8(s2,h,R0) - I8(s1,h,R0)
I8valnum   = hcubature(s -> I8integrand(s[1],h,R0) , (s1,), (s2,))[1]
I9val      = I9(s2,h,R0) - I9(s1,h,R0)
I9valnum   = hcubature(s -> I9integrand(s[1],h,R0) , (s1,), (s2,))[1]
I10val     = I10(s2,h,R0) - I10(s1,h,R0)
I10valnum  = hcubature(s -> I10integrand(s[1],h,R0) , (s1,), (s2,))[1]
I11val     = I11(s2,h,R0) - I11(s1,h,R0)
I11valnum  = hcubature(s -> I11integrand(s[1],h,R0) , (s1,), (s2,))[1]
I12val     = I12(s2,h,R0) - I12(s1,h,R0)
I12valnum  = hcubature(s -> I12integrand(s[1],h,R0) , (s1,), (s2,))[1]
I13val     = I13(s2,h,R0) - I13(s1,h,R0)
I13valnum  = hcubature(s -> I13integrand(s[1],h,R0) , (s1,), (s2,))[1]
I14hatval     = I14hat(s2,h,R0) - I14hat(s1,h,R0)
I14hatvalnum  = hcubature(s -> I14hatintegrand(s[1],h,R0) , (s1,), (s2,))[1]
I15hatval     = I15hat(s2,h,R0) - I15hat(s1,h,R0)
I15hatvalnum  = hcubature(s -> I15hatintegrand(s[1],h,R0) , (s1,), (s2,))[1]
I16hatval     = I16hat(s2,h,R0) - I16hat(s1,h,R0)
I16hatvalnum  = hcubature(s -> I16hatintegrand(s[1],h,R0) , (s1,), (s2,))[1]
I17hatval     = I17hat(s2,h,R0) - I17hat(s1,h,R0)
I17hatvalnum  = hcubature(s -> I17hatintegrand(s[1],h,R0) , (s1,), (s2,))[1]

@test I8val ≈ I8valnum 
@test I9val ≈ I9valnum 
@test I10val ≈ I10valnum 
@test I11val ≈ I11valnum
@test I12val ≈ I12valnum 
@test I13val ≈ I13valnum
@test I14hatval ≈ I14hatvalnum
@test I15hatval ≈ I15hatvalnum
@test I16hatval ≈ I16hatvalnum
@test I17hatval ≈ I17hatvalnum

Test Passed

## Appendix: Function Like Objects

### A tutorial Example Using Polynomials 
See [Function-like-objects in the Julia Manual](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects).

In [63]:
# define polynomial as vector of coefficients
struct Polynomial{Float64}
    coeffs::Vector{Float64}
end

# add function to evaluate the polynomial to the type definition   
function (p::Polynomial)(x)
    # use data stored in member function 
    v = p.coeffs[end]
    for i = (length(p.coeffs)-1):-1:1
         v = v*x + p.coeffs[i]
    end
    return v
end

# define in which value the polynomial should be evaluated in case that no arguments are passed 
(p::Polynomial)() = p(0)

In [3]:
p = Polynomial([1.,10.,100.])

Polynomial{Float64}([1.0, 10.0, 100.0])

In [4]:
dump(p)

Polynomial{Float64}
  coeffs: Array{Float64}((3,)) [1.0, 10.0, 100.0]


In [5]:
p(2.)

421.0

In [6]:
function do_integrate(poly)
    value = hcubature(x -> poly(x[1]) , (0,), (1,)) 
    return value
end 

do_integrate (generic function with 1 method)